README.md
ppmck
=====

BouKiCHi氏のppmck ex版のforkです。

もともとのREADME.mdはREADME-ex.mdへとリネームしてあります。


changelog.txt
2015-02-11 release 9a ex11.3 am1

	- #LINEを追加。
	- nes_include以下のドライバソースをリファクタリングした
	  (出力バイナリは変化なし)。

2014-08-07 release 9a ex11.3

improved effects in repeat2

2014-08-07 release 9a ex11.2

improved write timing of soft env.

2014-08-05 release 9a ex11.1

kコマンド使用時にも未処理のエフェクトを書き出すようにした。

2014-07-31 release 9a ex11

wコマンド使用時にそれまでのエフェクトの変化を書き出すようにした。

2014-07-23 release 9a ex10

@@、@@rをFME7で有効にした
readme.txtにあった変更点をchangelog.txtへ移動した

2014-04-13 release 9a ex9

#DEFINEを追加

2014-03-24 release 9a ex8.1

各効果コマンド書き出しタイミングをノート直前に変更
オーバーロード時の処理を修正


2014-03-22 release 9a ex8

@qの拡張とポルタメント記号の追加

2014-03-21 release 9a ex 7.3 

ノートオン直前のコマンド書き出し時にポインタを進めていたバグ修正

2014-02-26 release 9a ex 7.1

タイムシフト機能を最初のコマンドのみ有効にした。
スラーの計算を変更。

2014/02/24  release 9a ex 7

タイムシフト機能の搭載。

2013/09/06 release 9a ex 6.3

mknsf.sh/mknes.shの挙動を修正。

2013/08/21 release 9a ex 6.2

VRC7の@OTでDTパラメータがキャリアとモジュレータで反転していたのを修正。
changelog.txtを作成。


2013/08/10 release 9a ex 6.1

#EFFECT-BANKSWITCHの追加。


2013/07/10 release 9a ex6

OS X SnowLeopard向けのバイナリを添付。
過負荷処理の内容を変更。


2013/07/10 release 9a ex5.3

NSF再生時の過負荷の検出と抑制を実装した。



2013/06/26 release 9a ex5.2

nes_include以下のソースコードをUTF-8に変更した。
スラーを同一音階で使用する際に音長を長くするようにした。
VRC7の周波数テーブルを1/2に設定、エンベロープ等使用時のノイズを減少。


2013/05/08 release 9a ex5.1

コマンドライン系ソースコードと文章の文字コードをSJISからUTF-8に変更。
（win32コンソール向けにはgccオプションで内部的にSJISに変換）
nesasmのソースコードの調整。


2013/04/30 release 9a ex5

VRC7音色定義に@OTを追加、フォーマットはMGSDRVの音色定義に準拠。


2013/03/24 release 9a ex4

定義データが一定以上ある場合(1024バイト以上)、新しいバンクを利用するようにした。


2012/12/25 release 9a ex3

実機再生の為にFDS初期化の追加。


2010/03/05 release 9a ex2

mmc5のハードウェアエンベロープ機能を修正。
スラーが未対応のトラックではタイになるよう修正。


2010/01/01 release 9a ex1

release 9aの主な変更点であるppmck.asmをマージした。


2009/11/28 release 9 ex5.1

nesasmのバグ修正。srcディレクトリ内Makefileの調節。
sounddrv.hの長過ぎるラベルを32文字以内に修正。


2009/10/20 release 9 ex5

ハードウェアエンベロープ(EH)をノイズトラック、MMC5でも利用可能にした。


2009/05/13 release 9 ex4

DPCMでスラーコマンド後の最初のノートを発音しないようにした
当ファイルの訂正（スラーの有効トラックが異なった）
ppmckcのトラック有効フラグを調節
MMC5に対してスラーコマンドを有効にした


2009/05/13 release 9 ex3.5

当ファイルを改訂
DPCMトラックでスラーコマンドを無視するようにした


2008/03/19 release 9 ex3

2A03ハードウェアエンベロープの追加


2007/05/12 release 9 ex2

PSコマンドの追加


2007/05/04 release 9 ex

コンパイラにSM,SMOFコマンド、スラーコマンド後のノートチェックを追加

2007/05/03 release 9++ 

スラーの解釈を間違えていたという問題を修正（根本的）


2007/05/02 release 9+



src/nesasm/defs.h

/* path separator */
#if defined(DJGPP) || defined(MSDOS) || defined(WIN32)
#define PATH_SEPARATOR '\\'
#define PATH_SEPARATOR_STRING "\\"
#else
#define PATH_SEPARATOR '/'
#define PATH_SEPARATOR_STRING "/"
#endif

/* input file params */
#define MAX_FILENAME_LEN	115
#define MAX_PATH_LEN		2047
#define MAX_INCS			7
#define MAX_INCDIRS			10
#define INC_SEPARATOR		';'

/* machine */
#define MACHINE_PCE	0
#define MACHINE_NES	1

/* version string */
#define VERSION_STRING		"v2.51am1"

/* bank */
#define BANK_SIZE_SHIFT		13
#define BANK_SIZE			(1 << BANK_SIZE_SHIFT)
#define BANK_SIZE_MASK		(BANK_SIZE-1)

#define MAX_REAL_BANKS		128
#define MAX_PSEUDO_BANKS	128
#define BANK_MAP_SIZE		(MAX_REAL_BANKS+MAX_PSEUDO_BANKS)

#define RESERVED_BANK		(MAX_REAL_BANKS+0)
#define PROC_BANK			(MAX_REAL_BANKS+1)
#define GROUP_BANK			(MAX_REAL_BANKS+2)


#define MAX_BANK_NAME_LEN	63

/* bank:offset to linear address */
#define B2ADDR(b, o)		(((b)<<BANK_SIZE_SHIFT) + ((o)&BANK_SIZE_MASK))
#define ADDR2BNUM(a)		((a)>>BANK_SIZE_SHIFT)
#define ADDR2BOFS(a)		((a)&BANK_SIZE_MASK)

/* tile format for encoder */
#define CHUNKY_TILE		1
#define PACKED_TILE		2

/*
 * line buffer constraints
 */
/* max line length of source file */
#define MAX_LINE_CHARS		135
/* fine number field */
#define LFIELD				0
#define LFIELD_LEN			5
/* address field */
#define AFIELD				7
#define AFIELD_LEN			7
/* value field */
#define VFIELD				16
#define VFIELD_LEN			4
/* source field */
#define SFIELD				26
#define SFIELD_LEN			MAX_LINE_CHARS
/* line buffer size */
#define LINE_BUFFER_SIZE	(SFIELD+SFIELD_LEN+1)

/* symbol length */
#define SBOLSZ			32

/* common hash constraints */
#define COMMON_HASH_TABLE_SIZE	256
#define COMMON_HASH_MOD(v)	((v) & 0xFF)

/* macro argument types */
#define NO_ARG			0
#define ARG_REG			1
#define ARG_IMM			2
#define ARG_ABS			3
#define ARG_INDIRECT	4
#define ARG_STRING		5
#define ARG_LABEL		6

/* section types */
#define S_ZP			0
#define S_BSS			1
#define S_CODE			2
#define S_DATA			3
#define NUM_SECTIONS	4

/* assembler options */
#define OPT_LIST	 0
#define OPT_MACRO	 1
#define OPT_WARNING	 2
#define OPT_OPTIMIZE 3
#define NUM_OPTS	 4

/* assembler directives */
#define P_DB		 0	// .db
#define P_DW		 1	// .dw
#define P_DS		 2	// .ds
#define P_EQU		 3	// .equ
#define P_ORG		 4	// .org
#define P_PAGE		 5	// .page
#define P_BANK		 6	// .bank
#define P_INCBIN	 7	// .incbin
#define P_INCLUDE	 8	// .include
#define P_INCCHR	 9	// .incchr
#define P_INCSPR	10	// .incspr
#define P_INCPAL	11	// .incpal
#define P_INCBAT	12	// .incbat
#define P_MACRO		13	// .macro
#define P_ENDM		14	// .endm
#define P_LIST		15	// .list
#define P_MLIST		16	// .mlist
#define P_NOLIST	17	// .nolist
#define P_NOMLIST	18	// .nomlist
#define P_RSSET		19	// .rsset
#define P_RS		20	// .rs
#define P_IF		21	// .if
#define P_ELSE		22	// .else
#define P_ENDIF		23	// .endif
#define P_FAIL		24	// .fail
#define P_ZP		25	// .zp
#define P_BSS		26	// .bss
#define P_CODE		27	// .code
#define P_DATA		28	// .data
#define P_DEFCHR	29	// .defchr
#define P_FUNC		30	// .func
#define P_IFDEF		31	// .ifdef
#define P_IFNDEF	32	// .ifndef
#define P_VRAM		33	// .vram
#define P_PAL		34	// .pal
#define P_DEFPAL	35	// .defpal
#define P_DEFSPR	36	// .defspr
#define P_INESPRG	37	// .inesprg
#define P_INESCHR	38	// .ineschr
#define P_INESMAP	39	// .inesmap
#define P_INESMIR	40	// .inesmir
#define P_OPT		41	// .opt
#define P_INCTILE	42	// .inctile
#define P_INCMAP	43	// .incmap
#define P_MML		44	// .mml
#define P_PROC		45	// .proc
#define P_ENDP		46	// .endp
#define P_PGROUP	47	// .procgroup
#define P_ENDPG		48	// .endprocgroup
#define P_CALL		49	// .call

/* symbol flags */
#define MDEF	3	/* multiply defined */
#define UNDEF	1	/* undefined - may be zero page */
#define IFUNDEF	2	/* declared in a .if expression */
#define DEFABS	4	/* defined - two byte address */
#define MACRO	5	/* used for a macro name */
#define FUNC	6	/* used for a function */

/* operation code flags */
#define CLASSMASK	0x1FF8000
#define PSEUDO		0x0008000
#define CLASS1		0x0010000
#define CLASS2		0x0020000
#define CLASS3		0x0040000
#define CLASS5		0x0080000
#define CLASS6		0x0100000
#define CLASS7		0x0200000
#define CLASS8		0x0400000
#define CLASS9		0x0800000
#define CLASS10		0x1000000
#define MODEMASK	0x0001FFF	/* note: popcount(MODEMASK) == 13 */
#define ACC			0x0000001
#define IMM			0x0000002
#define ZP			0x0000004
#define ZP_X		0x0000008
#define ZP_Y		0x0000010
#define ZP_IND		0x0000020
#define ZP_IND_X	0x0000040
#define ZP_IND_Y	0x0000080
#define ABS			0x0000100
#define ABS_X		0x0000200
#define ABS_Y		0x0000400
#define ABS_IND		0x0000800
#define ABS_IND_X	0x0001000
#define MAX_ADDRMODES	16	/* for alignment, most least 2^n value
							 * more than popcount(MODEMASK) */
#define NUM_OPTYPES		6	/* number of optypes for class4 instructions */

/* pass flags */
#define FIRST_PASS	0
#define LAST_PASS	1

/* macro */
#define MAX_MACRO_ARGS		9
#define MAX_MACRO_DEPTHS	8
#define MAX_MACRO_NAME_LEN	31
#define MAX_MACRO_ARG_LEN	79

/* expr and func */
#define NUM_OP_STACK_DEPTH		64
#define NUM_VAL_STACK_DEPTH		64
#define NUM_FUNC_STACK_DEPTH	8
#define NUM_EXPR_STACK_DEPTH	NUM_FUNC_STACK_DEPTH*2
#define MAX_FUNC_ARGS			('9'-'1'+1)	/* assumes ASCII */
#define MAX_FUNC_ARG_LEN		79

/* structs */
typedef struct t_opcode {
	struct t_opcode *next;
	char  *name;
	void (*proc)(int *);
	int    flag;
	int    value;
	int    type_idx;
} t_opcode;

typedef struct t_input_info {
	FILE *fp;
	int   lnum;
	int   if_level;
	char  name[MAX_FILENAME_LEN+1];
} t_input_info;

typedef struct t_proc {
	struct t_proc *next;
	struct t_proc *link;
	struct t_proc *group;
	int  old_bank;
	int  bank;
	int  org;
	int  base;
	int  size;
	int  call;
	int  type;
	int  refcnt;
	char name[SBOLSZ+1];
} t_proc;

typedef struct t_symbol {
	struct t_symbol *next;
	struct t_symbol *local;
	struct t_proc   *proc;
	int  type;
	int  value;
	int  bank;
	int  page;
	int  nb;
	int  size;
	int  vram;
	int  pal;
	int  refcnt;
	int  reserved;
	int  data_type;
	int  data_size;
	char name[SBOLSZ+1];
} t_symbol;

typedef struct t_line {
	struct t_line *next;
	char *data;
} t_line;

typedef struct t_macro {
	struct t_macro *next;
	struct t_line *line;
	char name[SBOLSZ+1];
} t_macro;

typedef struct t_func {
	struct t_func *next;
	char line[MAX_LINE_CHARS+1];
	char name[SBOLSZ+1];
} t_func;

typedef struct t_tile {
	struct t_tile *next;
	unsigned char *data;
	unsigned int   crc;
	int index;
} t_tile;

typedef struct t_machine {
	int type;
	char *asm_name;
	char *asm_title;
	char *rom_ext;
	char *include_env;
	unsigned int zp_limit;
	unsigned int ram_limit;
	unsigned int ram_base;
	unsigned int ram_page;
	unsigned int ram_bank;
	struct t_opcode *inst;
	struct t_opcode *pseudo_inst;
    int  (*pack_8x8_tile)(unsigned char *, void *, int, int);
    int  (*pack_16x16_tile)(unsigned char *, void *, int,  int);
    int  (*pack_16x16_sprite)(unsigned char *, void *, int,  int);
    void (*write_header)(FILE *, int);
} MACHINE;

/* memset macro */
#define MEMCLR(x) memset(x,0,sizeof(x))

/* convert macro value to string literal */
#define MACROVAL_TO_STR_(x)	#x
#define MACROVAL_TO_STR(x)	MACROVAL_TO_STR_(x)
